### PR TITLE
feat: create REST APIs from graphQL queries & mutations

### LINKED ISSUE #1151: [DRAFT] Create REST APIs from GraphQL Queries & Mutations
Currently, we only support a GraphQL protocol on the ingress side of things. With this change, we will support REST APIs.
The main idea is to expose pre-defined GraphQL queries as REST endpoints. For eg:

If we have a typical Graphql Query that looks as follows:
```graphql
query getUsers {
  users {
    name
    age
    address {
      city
      country
      street
    }
  }
}
```

We should annotate it to support REST APIs as follows

```graphql
query GetUsers @rest(path: "/users") {
  users {
    name
    age
    address {
      city
      country
      street
    }
  }
}
```
More examples:

**Request with a `POST` method**
  ```graphql
  mutation CreateUser($input: CreateUserInput!) @rest(method: "POST", path: "/users") {
    createUser(input: $input) {
      user {
        id
        name
        email
      }
    }
  }
  ```
  **NOTE:** The default parameter should remain `GET`

**Request with route params**
  ```graphql
  query GetUser($id: ID!) @rest(method: "GET", path: "/users/$id") {
    user(id: $id) {
      id
      name
      email
    }
  }
  ```
  **NOTE:** We should be able to infer the type of the `id` parameter based on graphQL schema.

**Request with query params**
  By default all params in the query are assumed to be query params, unless explicitly selected as a part of route param.
  ```graphql
  query GetUser($id: ID!) @rest(path: "/users") {
    user(id: $id) {
      id
      name
      email
    }
  }
  ```

**Request with query and route params**
  
  ```graphql
  query GetUserDetails($id: ID!, $includePosts: Boolean) @rest(method: "GET", path: "/users/$id") {
    userDetails(id: $id, includePosts: $includePosts) {
      id
      name
      email
      posts @include(if: $includePosts) {
        title
        content
      }
    }
  }
  ```

### PR DESCRIPTION
**Pull request by shashitnak**
**Summary:**  
This PR adds the feature for automatically creating the REST endpoints for graphQL queries and mutations based on the values provided in the `@rest` directive.

**Issue Reference(s):**  
Fixes #1151 

**Build & Testing:**

- [ ] I ran `cargo test` successfully.
- [ ] I have run `./lint.sh --mode=fix` to fix all linting issues raised by `./lint.sh --mode=check`.

**Checklist:**

- [ ] I have added relevant unit & integration tests.
- [ ] I have updated the [documentation] accordingly.
- [ ] I have performed a self-review of my code.
- [x] PR follows the naming convention of `<type>(<optional scope>): <title>`

[documentation]: https://github.com/tailcallhq/tailcallhq.github.io/tree/develop/docs

### CODE REVIEW

File: src/config/config_module.rs [hunk 1]
```diff
@@ -11,6 +11,7 @@ use crate::config::Config;
 #[derive(Clone, Debug, Default, Setters)]
 pub struct ConfigModule {
     pub config: Config,
+    pub gql_requests: Arc<Vec<async_graphql::Request>>,
```
Comment (tusharmath) [inline src/config/config_module.rs:14]: can move this to extensions.

Comment (shashitnak) [reply src/config/config_module.rs]: I wasn't actually using this so I removed it

File: benches/impl_path_string_for_evaluation_context.rs [hunk 2]
```diff
@@ -186,6 +187,10 @@ impl<'a> ResolverContextLike<'a> for MockGraphqlContext {
     }
 
     fn add_error(&'a self, _: async_graphql::ServerError) {}
+
+    fn data<D: Any + Send + Sync>(&'a self) -> Option<&'a D> {
```
Comment (amitksingh1490) [inline benches/impl_path_string_for_evaluation_context.rs:191]: It should be set in request context

File: examples/rest_args.graphql [hunk 3]
```diff
@@ -0,0 +1,13 @@
+query GetPost($id: Int!) @rest(path: "/posts/$id", method: "GET") {
+  post(id: $id) {
+    id
+    title
+  }
+}
+
+query GetUser($input: Int!) @rest(path: "/users/$input") {
```
Comment (amitksingh1490) [inline examples/rest_args.graphql:8]: think about creating post request i.e a mutation

File: src/config/rest.rs [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+use serde::de::Visitor;
+use serde::{de, ser, Deserialize, Deserializer, Serialize, Serializer};
+use serde_json::de::StrRead;
+
+use crate::async_graphql_hyper::GraphQLRequest;
+use crate::blueprint::OperationQuery;
+use crate::http::Method;
+
+#[derive(Clone, Debug, Default)]
+pub struct RestApis(pub BTreeMap<Rest, String>);
```
Comment (amitksingh1490) [inline src/config/rest.rs:19]:
```suggestion
pub struct RestApis(BTreeMap<Rest, String>);
```

File: src/config/rest.rs [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+
+impl RestApis {
+    pub fn create_operations(&self) -> Vec<OperationQuery> {
```
Comment (amitksingh1490) [inline src/config/rest.rs:22]: define from

File: src/http/request_handler.rs [hunk 6]
```diff
@@ -128,8 +156,22 @@ pub async fn handle_request<T: DeserializeOwned + GraphQLRequestLike>(
...
         }
-
-        hyper::Method::GET if app_ctx.blueprint.server.enable_graphiql => graphiql(&req),
+        hyper::Method::GET
+            if req.uri().path() == "/" && app_ctx.blueprint.server.enable_graphiql =>
+        {
+            graphiql(&req)
+        }
+        ref method if req.uri().path().starts_with("/api") => {
+            let path = req.uri().path().strip_prefix("/api").unwrap();
```
Comment (amitksingh1490) [inline src/http/request_handler.rs:165]: get rid of unwrap

**Pull request closed without merging**